packages/integrations/vercel/src/serverless/middleware.ts

```diff
@@ -0,0 +1,58 @@
+import { fileURLToPath, pathToFileURL } from 'node:url';
+import esbuild from 'esbuild';
+import { join } from 'node:path';
+import { writeFile } from '../lib/fs.js';
+
+/**
+ * It generates the Vercel Edge Middleware file.
+ *
+ * It creates a temporary file, the edge middleware, with some dynamic info.
+ *
+ * Then this file gets bundled with esbuild. The bundle phase will inline the Astro middleware code.
+ *
+ * @param astroMiddlewareEntryPoint
+ * @param outPath
+ * @returns {Promise<URL>} The path to the bundled file
+ */
+export async function generateEdgeMiddleware(
+	astroMiddlewareEntryPoint: URL,
+	outPath: string
+): Promise<URL> {
```

Comment:
Is it correct that we're generating an edge middleware in [Vercel Serverless](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes) mode? Shouldn't the middleware run in the serverless runtime instead?

Reply:
Yes, that's correct. That's the intended logic. The middleware is an **edge** function, and it doesn't run in a serverless enviroment: https://vercel.com/docs/concepts/functions/edge-middleware#

From what I learnt, in Vercel an edge function is a Cloudflare worker, while a serverless function is an AWS lambda. Two different runtimes

Reply:
This is an opt-in feature.

The point of this feature is to run your middleware closer to the user. This means if you do something like a redirect to a language-specific subpath you do it close to the user. Then your regular app runs in serverless.

There wouldn't be a reason to have this feature for `@astrojs/vercel/edge`. You're already at the end. Splitting middleware into a separate function just makes it slower.

Reply:
If it's opt-in then that's good by me then 👍 I thought it will be a feature automatically turned on.

Reply:
> If it's opt-in then that's good by me then 👍 I thought it will be a feature automatically turned on.

Yeah, that's the whole philosophy. The user **must** set the correct configuration in order to trigger the feature. I reverted the code in the test case, Matthew caught the bug.